Clash doesn't build on GHC 9.6.7 (but does on 9.6.6)

I get:

```
src/Clash/Unique.hs:73:17: error: [GHC-83865]
    • Couldn't match type ‘Word64’ with ‘Int’
      Expected: GHC.Unique -> Unique
        Actual: GHC.Unique -> Word64
    • In the expression: GHC.getKey
      In an equation for ‘fromGhcUnique’: fromGhcUnique = GHC.getKey
   |
73 | fromGhcUnique = GHC.getKey
   |                 ^^^^^^^^^^
```

Perhaps:

```diff
diff --git a/clash-lib/src/Clash/Unique.hs b/clash-lib/src/Clash/Unique.hs
index 407b9ff64..551f3a48e 100644
--- a/clash-lib/src/Clash/Unique.hs
+++ b/clash-lib/src/Clash/Unique.hs
@@ -17,8 +17,18 @@ module Clash.Unique
   , fromGhcUnique
   ) where
 
-import Data.Word (Word64)
 #if MIN_VERSION_ghc(9,8,4)
+#define UNIQUE_IS_WORD64
+#endif
+
+#if MIN_VERSION_ghc(9,6,7)
+#if !MIN_VERSION_ghc(9,8,0)
+#define UNIQUE_IS_WORD64
+#endif
+#endif
+
+import Data.Word (Word64)
+#ifdef UNIQUE_IS_WORD64
 import GHC.Word (Word64(W64#))
 import GHC.Exts (Word64#)
 #else
@@ -31,7 +41,7 @@ import qualified GHC.Types.Unique as GHC
 import qualified Unique as GHC
 #endif
 
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 type Unique = Word64
 type Unique# = Word64#
 
@@ -63,7 +73,7 @@ instance Uniquable Unique where
   getUnique = id
   setUnique = flip const
 
-#if !MIN_VERSION_ghc(9,8,4)
+#ifndef UNIQUE_IS_WORD64
 instance Uniquable Word64 where
   getUnique = fromIntegral
   setUnique _ = fromIntegral
```

--- a/.ci/docker/build-and-publish-docker-image.sh
+++ b/.ci/docker/build-and-publish-docker-image.sh
@@ -14,7 +14,7 @@ elif [[ "$1" != "" ]]; then
   exit 1
 fi
 UBUNTU_VERSION=jammy-20250126
-GHC_VERSIONS=("9.10.1" "9.8.4"  "9.6.6"  "9.4.8"  "9.2.8"   "9.0.2"   "8.10.7")
+GHC_VERSIONS=("9.10.1" "9.8.4"  "9.6.7"  "9.4.8"  "9.2.8"   "9.0.2"   "8.10.7")
 CABAL_VERSIONS=("3.14.1.1" "3.14.1.1" "3.14.1.1" "3.14.1.1" "3.14.1.1" "3.14.1.1" "3.12.1.0")
 
 # We want to use docker buildkit so that our layers are built in parallel. This

--- a/.ci/gitlab/benchmark.yml
+++ b/.ci/gitlab/benchmark.yml
@@ -1,5 +1,5 @@
 .benchmark:
-  image: ghcr.io/clash-lang/clash-ci:$GHC_VERSION-20250305
+  image: ghcr.io/clash-lang/clash-ci:$GHC_VERSION-20250402
   stage: test
   timeout: 2 hours
   variables:

--- a/.ci/gitlab/common.yml
+++ b/.ci/gitlab/common.yml
@@ -12,7 +12,7 @@ default:
   timeout: 10 minutes
   stage: build
   variables:
-    CLASH_DOCKER_TAG: 20250305
+    CLASH_DOCKER_TAG: 20250402
     CACHE_BUST_TOKEN: 3
     # Note that we copy+paste the image name into CACHE_FALLBACK_KEY. If we don't,
     # $GHC_VERSION gets inserted at verbatim, instead of resolving to some ghc version.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -145,7 +145,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        ghc: [ "8.10.7", "9.0.2", "9.6.6" ]
+        ghc: [ "8.10.7", "9.0.2", "9.6.7" ]
         include:
           - multiple_hidden: yes
 
@@ -159,7 +159,7 @@ jobs:
 
     # Run steps inside the clash CI docker image
     container:
-      image: ghcr.io/clash-lang/clash-ci:${{ matrix.ghc }}-20250305
+      image: ghcr.io/clash-lang/clash-ci:${{ matrix.ghc }}-20250402
 
       env:
         THREADS: 2

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -43,7 +43,7 @@ tests:
         SKIP_CLASH_FFI_EXAMPLE: "yes"
         RUN_TESTS: "always"
 
-      - GHC_VERSION: 9.6.6
+      - GHC_VERSION: 9.6.7
         RUN_TESTS: "always"
 
       - GHC_VERSION: [9.4.8, 9.2.8]

--- a/changelog/2025-04-02T09_08_44+02_00_add_ghc967_support
+++ b/changelog/2025-04-02T09_08_44+02_00_add_ghc967_support
@@ -1,0 +1,1 @@
+FIXED: Clash did not build on GHC 9.6.7 (but did on 9.6.6) [#2916](https://github.com/clash-lang/clash-compiler/issues/2916)

--- a/clash-lib-hedgehog/src/Clash/Hedgehog/Unique.hs
+++ b/clash-lib-hedgehog/src/Clash/Hedgehog/Unique.hs
@@ -18,6 +18,10 @@ module Clash.Hedgehog.Unique
   , sampleUniqMapBiased
   ) where
 
+#if MIN_VERSION_ghc(9,8,4) || (MIN_VERSION_ghc(9,6,7) && !MIN_VERSION_ghc(9,8,0))
+#define UNIQUE_IS_WORD64
+#endif
+
 import Control.Applicative (Alternative(empty))
 import Data.Either (rights)
 import Hedgehog (MonadGen, Range)
@@ -35,7 +39,7 @@ import Clash.Hedgehog.Internal.Bias
 
 genUnique :: forall m. MonadGen m => m Unique
 genUnique =
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
   Gen.word64
 #else
   Gen.int

--- a/clash-lib/src/Clash/Core/FreeVars.hs
+++ b/clash-lib/src/Clash/Core/FreeVars.hs
@@ -29,11 +29,15 @@ module Clash.Core.FreeVars
   )
 where
 
+#if MIN_VERSION_ghc(9,8,4) || (MIN_VERSION_ghc(9,6,7) && !MIN_VERSION_ghc(9,8,0))
+#define UNIQUE_IS_WORD64
+#endif
+
 import qualified Control.Lens           as Lens
 import Control.Lens.Fold                (Fold)
 import Control.Lens.Getter              (Contravariant)
 import Data.Coerce
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 import qualified GHC.Data.Word64Set     as IntSet
 #else
 import qualified Data.IntSet            as IntSet
@@ -86,7 +90,7 @@ typeFreeVars'
   :: (Contravariant f, Applicative f)
   => (forall b . Var b -> Bool)
   -- ^ Predicate telling whether a variable is interesting
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
   -> IntSet.Word64Set
 #else
   -> IntSet.IntSet

--- a/clash-lib/src/Clash/Core/VarEnv.hs
+++ b/clash-lib/src/Clash/Core/VarEnv.hs
@@ -97,13 +97,17 @@ module Clash.Core.VarEnv
   )
 where
 
+#if MIN_VERSION_ghc(9,8,4) || (MIN_VERSION_ghc(9,6,7) && !MIN_VERSION_ghc(9,8,0))
+#define UNIQUE_IS_WORD64
+#endif
+
 import           Control.DeepSeq           (NFData)
 import           Data.Binary               (Binary)
 import           Data.Coerce               (coerce)
 import qualified Data.List                 as List
 import qualified Data.List.Extra           as List
 import           Data.Maybe                (fromMaybe)
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 import           Data.Word                 (Word64)
 #endif
 
@@ -389,7 +393,7 @@ eltsVarSet = UniqMap.elems
 -- * InScopeSet
 
 type Seed
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
   = Word64
 #else
   = Int

--- a/clash-lib/src/Clash/Data/UniqMap.hs
+++ b/clash-lib/src/Clash/Data/UniqMap.hs
@@ -39,13 +39,17 @@ module Clash.Data.UniqMap
   , elems
   ) where
 
+#if MIN_VERSION_ghc(9,8,4) || (MIN_VERSION_ghc(9,6,7) && !MIN_VERSION_ghc(9,8,0))
+#define UNIQUE_IS_WORD64
+#endif
+
 import           Prelude hiding (elem, filter, lookup, notElem, null)
 
 import           Control.DeepSeq (NFData)
 import           Data.Binary (Binary (..))
 import           Data.Bifunctor (first)
 import           Data.Function (on)
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 import           GHC.Data.Word64Map.Strict (Word64Map)
 import qualified GHC.Data.Word64Map.Strict as IntMap
 #else
@@ -71,7 +75,7 @@ import           Clash.Unique (Unique, Uniquable(getUnique))
 -- uniqueable and provide their own key, however a unique can be associated
 -- with any value.
 newtype UniqMap a
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
   = UniqMap { uniqMapToIntMap :: Word64Map a }
 #else
   = UniqMap { uniqMapToIntMap :: IntMap a }
@@ -85,7 +89,7 @@ newtype UniqMap a
     , Semigroup
     , Show
     )
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 instance Binary a => Binary (UniqMap a) where
   put (UniqMap m) = put (IntMap.size m) <> mapM_ put (IntMap.toAscList m)
   get             = fmap (UniqMap . IntMap.fromDistinctAscList) get

--- a/clash-lib/src/Clash/Unique.hs
+++ b/clash-lib/src/Clash/Unique.hs
@@ -17,8 +17,13 @@ module Clash.Unique
   , fromGhcUnique
   ) where
 
+#if MIN_VERSION_ghc(9,8,4) || (MIN_VERSION_ghc(9,6,7) && !MIN_VERSION_ghc(9,8,0))
+#define UNIQUE_IS_WORD64
+#endif
+
 import Data.Word (Word64)
-#if MIN_VERSION_ghc(9,8,4)
+
+#ifdef UNIQUE_IS_WORD64
 import GHC.Word (Word64(W64#))
 import GHC.Exts (Word64#)
 #else
@@ -31,7 +36,7 @@ import qualified GHC.Types.Unique as GHC
 import qualified Unique as GHC
 #endif
 
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 type Unique = Word64
 type Unique# = Word64#
 
@@ -63,7 +68,7 @@ instance Uniquable Unique where
   getUnique = id
   setUnique = flip const
 
-#if !MIN_VERSION_ghc(9,8,4)
+#ifndef UNIQUE_IS_WORD64
 instance Uniquable Word64 where
   getUnique = fromIntegral
   setUnique _ = fromIntegral

--- a/clash-lib/src/Clash/Util/Graph.hs
+++ b/clash-lib/src/Clash/Util/Graph.hs
@@ -14,9 +14,13 @@ module Clash.Util.Graph
   , callGraphBindings
   ) where
 
+#if MIN_VERSION_ghc(9,8,4) || (MIN_VERSION_ghc(9,6,7) && !MIN_VERSION_ghc(9,8,0))
+#define UNIQUE_IS_WORD64
+#endif
+
 import           Data.Tuple            (swap)
 import           Data.Foldable         (foldlM)
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 import qualified GHC.Data.Word64Map.Strict as IntMap
 import qualified GHC.Data.Word64Set        as IntSet
 #else
@@ -33,7 +37,7 @@ import           Clash.Driver.Types (BindingMap, Binding (bindingTerm))
 import           Clash.Normalize.Util (callGraph)
 import           Clash.Unique (Unique)
 
-#if MIN_VERSION_ghc(9,8,4)
+#ifdef UNIQUE_IS_WORD64
 type IntMap = IntMap.Word64Map
 type IntSet = IntSet.Word64Set
 #endif

